Sources/Tests/CoreTests/test_endianSwap.cpp

```diff
@@ -0,0 +1,146 @@
+/*==LICENSE==*
+
+CyanWorlds.com Engine - MMOG client, server and tools
+Copyright (C) 2011  Cyan Worlds, Inc.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include <gtest/gtest.h>
+
+#include "HeadSpin.h"
+
+typedef union {
+    uint16_t i;
+    uint8_t c[2];
+} endian16_t;
+
+typedef union {
+    uint32_t i;
+    uint8_t c[4];
+} endian32_t;
+
+typedef union {
+    uint64_t i;
+    uint8_t c[8];
+} endian64_t;
+
+typedef union {
+    float i;
+    uint8_t c[4];
+} endianFloat_t;
+
+typedef union {
+    double i;
+    uint8_t c[8];
+} endianDouble_t;
+
+
+TEST(endianSwap, detection_accuracy)
+{
+    endian32_t val = { 0x01020304 };
+
+#ifdef HS_BIG_ENDIAN
+    EXPECT_EQ(val.c[0], 0x01);
+#else
+    EXPECT_EQ(val.c[0], 0x04);
+#endif
```

Comment:
Unfortunately, this sort of type punning (even using a union) is UB in C++ (even though it's perfectly valid and legal in C).  It may work on supported compilers for now, but it's not guaranteed to in the future...  The "valid" ways to do this are either with memcpy or `std::bit_cast` (C++20).

Reply:
_grumble grumble_ Okay, I've moved everything to using memcpy